### PR TITLE
[SPARK-42217][SQL] Support implicit lateral column alias in queries with Window

### PR DESCRIPTION
**Pull request by anchovYu**
### What changes were proposed in this pull request?

The code is in a reviewable state.
Todo:
- [x] PR description, with explanation on unsupported cases
- [x] Negative test cases (make sure throw the same exception as without LCA) 

This PR adds support for using implicit lateral column alias in queries with Window, for example:
```sql
create table employee(dept integer, name string, salary integer, bonus integer, properties struct<joinYear integer, mostRecentEmployer string>)

-- Contrived example: LCA in Project queries with Window
-- Below covers: LCA used in Window expressions, Window expressions as LCA, chaining
select
  name as n,
  dept as d, 
  rank() over (partition by d order by salary) as rank, 
  min(salary) over (partition by rank order by d)
from employee

-- Contrived example: LCA in Aggregate queries with Window
-- Below covers: LCA used in Window expression, Window expressions as LCA, chaining 
select
  dept as d,
  properties.joinYear as jy, 
  sum(avg(bonus)) over (partition by jy order by d) as sum_avg, 
  min(jy) over (partition by sum_avg order by dept) 
from employee
group by dept, properties.joinYear
```
Generally it provides full LCA support for Window queries. 
**Update: for performance concern as stated in the comments, I updated the PR to ban all the LCA usage in Window expression.** We can further iterate and discuss if we want to allow it in later PRs, if there are requests. Currently these queries fail with `UNSUPPORTED_FEATURE.LATERAL_COLUMN_ALIAS_IN_WINDOW`.

The following case is not supported:
* LCA in Aggregate queries with Window, with Having clause.
  It throws an `UNSUPPORTED_FEATURE.LATERAL_COLUMN_ALIAS_IN_AGGREGATE_WITH_WINDOW_AND_HAVING` exception.
  For example,
  ```sql
  > select dept as d, rank() over (partition by d order by avg(salary)) as rank
    from employee 
    group by dept
    having sum(bonus) < 3000

  org.apache.spark.sql.AnalysisException: [UNSUPPORTED_FEATURE.LATERAL_COLUMN_ALIAS_IN_AGGREGATE_WITH_WINDOW_AND_HAVING]
  The feature is not supported: Referencing lateral column alias `d` in the aggregate query both with window
  expressions and with having clause. Please rewrite the aggregate query by removing the having clause or
  removing lateral alias reference in the SELECT list.
  ```

Note that Spark initially don't have full support for Aggregate queries with Window, with Order by clause (sort). For example, the following query **without LCA**:
```sql
select dept, sum(sum(salary)) over (partition by dept order by sum(salary)) as sum_sum
from employee
group by dept
order by sum(bonus)

org.apache.spark.sql.AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `bonus` cannot be resolved.
```
Therefore LCA doesn't have the support for the above query type as well.

#### Implementation Details
The implementation doesn't really change a lot in the rules itself, but to change the rule application order. This PR changes to apply the rule `ResolveLateralColumnAliasReference` before `ExtractWindowExpressions`.
That is, it guarantees when it starts to add Window (`ExtractWindowExpressions`), there is no `LateralColumnAliasReference` left in the Project and Aggregate. They are already handled and removed by `ResolveLateralColumnAliasReference`.

For example, given a query plan:
```
!'Project [name#6 AS n#52, dept#5 AS d#53, rank() windowspecdefinition(lateralAliasReference(d#53, d, d#53), salary#7 ASC NULLS FIRST, specifiedwindowframe(..) AS rank#54]
!+- SubqueryAlias spark_catalog.default.employee  
!   +- Relation spark_catalog.default.employee[dept#5,name#6,salary#7,bonus#8,properties#9]                                                                                             

1. First apply rule ResolveLateralColumnAliasReference, adding a Project below:
!'Project [name#6 AS n#52, d#53, rank() windowspecdefinition(d#53, salary#7 ASC NULLS FIRST, specifiedwindowframe(..) AS rank#54]
 +- Project [dept#5, name#6, salary#7, bonus#8, properties#9, dept#5 AS d#53]    
    +- SubqueryAlias spark_catalog.default.employee  
       +- Relation spark_catalog.default.employee[dept#5,name#6,salary#7,bonus#8,properties#9] 

2. Second apply rule ExtractWindowExpressions, adding a Window above:
Window [name#6 AS n#52, d#53, rank(salary#7) windowspecdefinition(d#53, salary#7 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rank#54], [d#53], [salary#7 ASC NULLS FIRST]
+- Project [name#6, d#53, salary#7]
   +- Project [dept#5, name#6, salary#7, bonus#8, properties#9, dept#5 AS d#53]
      +- SubqueryAlias spark_catalog.default.employee
         +- Relation spark_catalog.default.employee[dept#5,name#6,salary#7,bonus#8,properties#9] orc
```
With this rule application order, we made the minimal code change to support the feature. It doesn't need to change the LCA rule to match the Window operator and do specific handing, with potential limitation.

### Why are the changes needed?

To provide full functionality for using LCA in the SELECT clauses.

### Does this PR introduce _any_ user-facing change?

Yes, new functionality supported. View the description.

### How was this patch tested?

Newly added tests.

### CODE REVIEW

**Comment by anchovYu** [comment]
@gengliangwang @cloud-fan @srielau

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 1]
```diff
@@ -81,6 +82,9 @@ import org.apache.spark.sql.internal.SQLConf
  *    +- Aggregate [dept#14] [avg(salary#16) AS avg(salary)#26, avg(bonus#17) AS avg(bonus)#27,
  *                            dept#14]
  *       +- Child [dept#14,name#15,salary#16,bonus#17]
+ *
+ *
+ * TODO: example with Window
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:87]: @anchovYu could you fix the TODO here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 2]
```diff
@@ -104,7 +112,10 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
       // phase 2: unwrap
       plan.resolveOperatorsUpWithPruning(
         _.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE), ruleId) {
-        case p @ Project(projectList, child) if p.resolved
+        case p @ Project(projectList, child) if
+          (p.resolved
+            // can also work on nodes that contains Window but have all others resolved
+            || (projectList.exists(hasWindowExpression) && exprsAndChildResolved(p)))
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:118]: `p.resolved || (projectList.exists(hasWindowExpression) && exprsAndChildResolved(p))` shows up multiple times. Shall we abstract them into a method?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 3]
```diff
@@ -17,7 +17,8 @@
 
 package org.apache.spark.sql.catalyst.analysis
 
-import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, AttributeMap, Expression, LateralColumnAliasReference, NamedExpression, ScalarSubquery}
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, AttributeMap, Expression, LateralColumnAliasReference, NamedExpression, ScalarSubquery, WindowExpression, WindowSpecDefinition}
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:20]: nit: let's just use `import org.apache.spark.sql.catalyst.expressions._`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala [hunk 4]
```diff
@@ -177,8 +195,18 @@ object ResolveLateralColumnAliasReference extends Rule[LogicalPlan] {
...
-          val projectExprs = aggregateExpressions.map { exp =>
-            exp.transformDown {
+          // Extract the expressions to keep in the Aggregate. Return the transformed expression
+          // fully substituted with the attribute reference to the extracted expressions.
+          def extractExpressions(expr: Expression): Expression = {
+            expr match {
+              case w @ WindowExpression(function, spec) =>
+                // Manually skip the handling on the function itself but iterate on its children
+                // instead. This is to avoid extracting a window expression's aggregate functions.
+                // For example, for `sum(sum(col1)) over (..)`, we should extract sum(col1) but not
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAliasReference.scala:205]: Maybe we should be more explicit: `WindowExpression.function` can be `AggregateExpression`, but we don't want to extract that `AggregateExpression`.

**Review by cloud-fan — APPROVED** [review]
LGTM, please address comments from @gengliangwang

**Review by gengliangwang — APPROVED** [review]

**Comment by srielau** [comment]
What happens when you chain? Do we treat OLAP functions like scalar function (they kind of are) and start stacking.
That can make a mess of performance

**Comment by anchovYu** [comment]
@srielau Thanks for bringing up this concern. 
LCA in this PR doesn't introduce nested Window, it is a feature already been supported by Spark. But LCA does make this pattern easier to access.

I suggest to merge this PR first, but further iterate on if we want to control the default behavior using a config, or if we want to partially forbid it. I believe basic support in Window is still needed anyway, e.g. select 1 as a, a, ..(followed by some  regular window expression), which is enabled by this PR.

**Comment by srielau** [comment]
> @srielau Thanks for bringing up this concern. LCA in this PR doesn't introduce nested Window, it is a feature already been supported by Spark. But LCA does make this pattern easier to access.
> 
> I suggest to merge this PR first, but further iterate on if we want to control the default behavior using a config, or if we want to partially forbid it. I believe basic support in Window is still needed anyway, e.g. select 1 as a, a, ..(followed by some regular window expression), which is enabled by this PR.

LGTM

**Comment by gengliangwang** [comment]
@anchovYu would you resolve the conflict? I will merge this one after the tests passed again

**Comment by anchovYu** [comment]
Thanks @gengliangwang !
@srielau , after second consideration and brief discussion w/ Wenchen, I updated the PR to ban all LCA usage in Window for extra safe. 'It's easier to relax the limit than to tighten it'.

**Comment by gengliangwang** [comment]
OK, the latest commit LGTM too

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.4 
cc @xinrong-meng

**Pull request closed without merging**